source/manual/intro-to-govuk-docker.html.md

```diff
@@ -0,0 +1,160 @@
+
+Images are downloaded from the Docker Registry. The containers are run as a process on our host.
+
+Another image is ruby:
+
+```shell
+docker run -it --rm ruby:2.6.3 bash
+```
+
+Here we get a ruby image and specifying which version we want. The `bash` at the end means we get a bash shell.
+
+### Volumes
+
+* Add a volume by running:
+
+```shell
+docker run -it --rm -v $(pwd):/app ruby:2.6.3 bash
+```
+
+This will map your current directory (being the root of the [content-publisher] project) to `/app` inside of the container.
+
+If you were to run bundle install for the app, the next time we spin up a container all of the gems would need re-installing.
+
+#### Persistent volumes
+
+Docker allows you create separate volumes for persistent data.
+
+* Create a persistent volume for our gems:
+
+```shell
+docker volume create content-publisher-bundle
+```
+
+*   Run docker with that volume:
+
+```shell
+docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle ruby:2.6.3 bash
+```
+
+*   Then installing gems:
+```shell
+cd /app
+bundle install
+```
+
+### Specifying dependencies (node, PostgreSQL, DNS)
+
+* Try to run the tests:
+
+```shell
+bundle exec rake
+```
+
+ We get an error about “could not find javascript runtime”. Our Ruby image isn’t doing quite what we need, so we’re going to create our own image based off that
+
+*   Create a Dockerfile in the content-publisher project:
+
+```
+
+    FROM ruby:2.6.3
+
+
+    RUN curl -sL https://deb.nodesource.com/setup_12.x | bash
+
+
+    RUN apt-get install nodejs && apt-get clean
+
+
+    RUN apt-get install -y libxss1 libappindicator1 libindicator7 && apt-get clean
+
+
+    RUN wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
+
+
+    RUN dpkg -i google-chrome-stable_current_amd64.deb; apt-get -fy install
+
+
+    RUN useradd -m build
+
+
+    USER build
+
+
+    ```
+
+*   `docker build -t content-publisher .`
+*   So now we can run it as: `docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle content-publisher bash`
+
+
+### Postgres image/container
+
+
+
+*   `docker run -it --rm postgres`
+*   `docker ps` - to see what containers you have running
+*   `docker inspect [container ID for postgres]`. Grab the IP address
+*   In the content-publisher container:
+    *   `export DATABASE_URL=postgres://postgres@172.17.0.3/content-publisher-dev`
+    *   `export TEST_DATABASE_URL=postgres://postgres@172.17.0.3/content-publisher-test`
+*   Run the tests in content-publisher container: `bundle exec rake db:setup`
+*   Create a volume to persist to for the data: `docker volume create content-publisher-postgres`
+*   `docker run -it --rm -v content-publisher-postgres:/var/lib/postgresql/data postgres`
+*   Add `builder` user to Dockerfile (see above) then re-build content-publisher container: `docker build -t content-publisher .`
```

Comment:
This looks like it's got mixed up with the Chrome stuff - appreciate I was going rather fast at this point :-|.